fog/ledger/server/src/db_fetcher.rs

```diff
@@ -0,0 +1,308 @@
+            next_block_index,
+            logger,
+            enclave,
+            watcher,
+            db_poll_shared_state,
+        };
+        thread.run();
+    }
+
+    fn run(mut self) {
+        log::info!(self.logger, "Db fetcher thread started.");
+        self.next_block_index = 0;
+        loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Db fetcher thread stop requested.");
+                break;
+            }
+            // Each call to load_block_data attempts to load one block for each known
+            // invocation. We want to keep loading blocks as long as we have data to load,
+            // but that could take some time which is why the loop is also gated
+            // on the stop trigger in case a stop is requested during loading.
+            while self.load_block_data() && !self.stop_requested.load(Ordering::SeqCst) {}
+
+            sleep(DB_POLL_INTERNAL);
+        }
+    }
+
+    /// Attempt to load the next block for each of the ledger db invocations we
+    /// are aware of and tracking.
+    /// Returns true if we might have more block data to load.
+    fn load_block_data(&mut self) -> bool {
+        let mut has_more_work = false;
+        let mut records: Vec<KeyImageData> = Vec::new();
+        let watcher_timeout: Duration = Duration::from_millis(5000);
+
+        match self.db.get_block_contents(self.next_block_index) {
+            Err(LedgerError::NotFound) => std::thread::sleep(Self::POLLING_FREQUENCY),
+            Err(e) => {
+                log::error!(
+                    self.logger,
+                    "Unexpected error when checking for block data {}: {:?}",
+                    self.next_block_index,
+                    e
+                );
+                std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+            }
+            Ok(block_contents) => {
+                // Get the timestamp for the block.
+                let timestamp = Self::get_watcher_timestamp(
+                    self.next_block_index,
+                    &self.watcher,
+                    watcher_timeout,
+                    &self.logger,
+                );
+
+                for key_image in block_contents.key_images {
+                    let mut rec = KeyImageData {
+                        key_image: KeyImage::from(2),
+                        block_index: u64::MAX,
+                        timestamp: u64::MAX,
+                    };
+
+                    rec.key_image = key_image;
+                    rec.block_index = self.next_block_index;
+                    rec.timestamp = timestamp;
+
+                    records.push(rec);
+                }
+
+                self.add_records_to_enclave(self.next_block_index, records);
+                let mut shared_state = self.db_poll_shared_state.lock().expect("mutex poisoned");
+                shared_state.highest_processed_block_count = self.next_block_index + 1;
```

Comment:
we should add a code comment like 
```
this is next_block_index + 1 because next_block_index is actually the block we just processed, so we have fully processed next_block_index + 1 blocks
```